### PR TITLE
tls: use Go default kex for the moment that include PQC

### PR DESCRIPTION
**Pull request by bwesterb**
By default Go 1.23 enables X25519Kyber768, a post-quantum key agreement method that is enabled by default on Chrome. Go 1.23 does not expose the CurveID, so we cannot add it by specifying it in CurvePreferences. The reason is that X25519Kyber768 is a preliminary key agreement that will be supplanted by X25519MLKEM768. For the moment there is value in enabling it.

A consequence of this is that by default Caddy will enable support for P-384 and P-521.

This PR also removes the special code to add support for X25519Kyber768 via the Cloudflare Go branch.

Cf #6540

### CODE REVIEW

**Review by sam-bee — COMMENTED** [review]
Is this having the desired effect? I seem to be still getting `X25519`, but I'm not sure why.

QA environment details:
```
root@5a81f85b51be:/go/src/github.com/caddyserver/caddy/cmd/caddy# git remote -v
origin	https://github.com/bwesterb/caddy.git (fetch)
origin	https://github.com/bwesterb/caddy.git (push)
root@5a81f85b51be:/go/src/github.com/caddyserver/caddy/cmd/caddy# git rev-parse --abbrev-ref HEAD
pq123
root@5a81f85b51be:/go/src/github.com/caddyserver/caddy/cmd/caddy# git log -1 --oneline
e93d68b (HEAD -> pq123, origin/pq123) tls: use Go default kex for the moment that include PQC
root@5a81f85b51be:/go/src/github.com/caddyserver/caddy/cmd/caddy# caddy --version
e93d68b61276d8706698cdfb9b4b2583a9c97a41 (26 Aug 24 22:21 UTC)
root@5a81f85b51be:/go/src/github.com/caddyserver/caddy/cmd/caddy# go version
go version go1.23.0 linux/amd64
```

File: modules/caddytls/values.go [hunk 1]
```diff
@@ -108,6 +108,11 @@ var supportedCertKeyTypes = map[string]certmagic.KeyType{
 // implementation exists (e.g. P256). The latter ones can be
 // found here:
 // https://github.com/golang/go/tree/master/src/crypto/elliptic
+//
+// Temporily we ignore these default, to take advantage of X25519Kyber768
+// in Go's defaults (X25519Kyber768, X25519, P-256,  P-384, P-521), which
```
Comment (francislavoie) [inline modules/caddytls/values.go:113]:
```suggestion
// in Go's defaults (X25519Kyber768, X25519, P-256, P-384, P-521), which
```

File: modules/caddytls/connpolicy.go [hunk 2]
```diff
@@ -841,7 +841,15 @@ func setDefaultTLSParams(cfg *tls.Config) {
...
 	if len(cfg.CurvePreferences) == 0 {
-		cfg.CurvePreferences = defaultCurves
+		// We would want to write
+		//
+		//	cfg.CurvePreferences = defaultCurves
+		//
+		// but that would disable the post-quantum key agreement X25519Kyber768
+		// supported in Go 1.23, for which the CurveID is not exported.
+		// Instead, we'll set CurvePreferences to nil, which will enable PQC.
+		// See https://github.com/caddyserver/caddy/issues/6540.
```
Comment (francislavoie) [inline modules/caddytls/connpolicy.go:851]: The dot breaks the link when viewing in github, for some reason it doesn't ignore it :man_shrugging: 

```suggestion
		// See https://github.com/caddyserver/caddy/issues/6540
```

**Comment by bwesterb** [comment]
> Is this having the desired effect? I seem to be still getting `X25519`, but I'm not sure why.

Can confirm. Debugging now.

**Comment by bwesterb** [comment]
Ok, I figured it out. Go disables new features controlled by debug flags when the `go.mod` go version is too low.

**Review by sam-bee — APPROVED** [review]
I have retested the changes, and this looks good to me.

Screenshot from 2024-08-27 14-08-59

_X25519Kyber768Draft00 ciphersuite, reported by Chrome_

QA env:
```
root@97e644d656f4:/go/src/github.com/caddyserver/caddy/cmd/caddy# git log -1 --oneline
9051820 (HEAD -> pq123, origin/pq123) tls: use Go default kex for the moment that include PQC
root@97e644d656f4:/go/src/github.com/caddyserver/caddy/cmd/caddy# caddy --version
9051820d85db00232c5ffb4403899ad4f812229d (27 Aug 24 11:21 UTC)
root@97e644d656f4:/go/src/github.com/caddyserver/caddy/cmd/caddy# go version
go version go1.23.0 linux/amd64
```
I take it there's nothing special that needs to happen if the Caddy user specifies TLS 1.2?

**Comment by bwesterb** [comment]
> Ok, I figured it out. Go disables new features controlled by debug flags when the `go.mod` go version is too low.

There is a workaround pointed out by @FiloSottile by adding a go:debug line in main.

**Review by mholt — APPROVED** [review]
Excellent, LGTM. Thanks for everyone's input!

(The lint failure is the result of a bug in the linter, hopefully they fix it soon.)

**Pull request merged**